### PR TITLE
ENH: Add a pinvh function to linalg.

### PR DESCRIPTION
**Pull request by jordan-melendez**
This is a faster version of `linalg.pinv` for Hermitian matrices. It is like the Scipy version of `pinvh`, but supports stacked matrices like Numpy's `pinv`.

### CODE REVIEW

**Comment by eric-wieser** [comment]
I think you should perhaps instead add a `hermitian=False` argument to pinv, like we do here:

https://github.com/numpy/numpy/blob/4abe9d8c85aae09eaa79212350c1e85efbe922e9/numpy/linalg/linalg.py#L1756

**Comment by rgommers** [comment]
Is there a reason to put this in NumPy rather than extend the SciPy version? In general we'd like to keep the two in sync for features we have already in both, and put new features in SciPy unless there's a really good reason to put things in NumPy.

Note that incompatible changes will also break (or not have) Numba support for those new features. Numba uses `scipy.linalg.cython_blas/lapack` when it sees `numpy.linalg` functions, so anything added that's not in SciPy will not be supported.

**Comment by jordan-melendez** [comment]
@eric-wieser that could work! I was just trying to mirror the format of Scipy here.

@rgommers My reasoning was that Scipy uses its own internal functions, like `eigh`, which only operate on (M,M) shaped arrays. From what I can tell, "2d matrices only" seems to be a common theme of Scipy's linalg library, whereas Numpy provides a lot of the same functionality, but allows matrices to be stacked. I assume Scipy has a good reason for this restriction, and I don't entirely know what it is or how I would modify it if I wanted to.

In regards to the Numba issue, what would it take to add that support? I'm not super familiar with the way it interacts with Numpy/Scipy and using BLAS/LAPACK directly.

Thanks!

**Comment by eric-wieser** [comment]
@rgommers: I think the fact we allowed a hermitian overload for `matrix_rank` suggests it would be reasonable to add one for the rest of linalg, at least in places where no new lapack functions are needed.

**Comment by eric-wieser** [comment]
You could save a lot of duplication here by adding an overload as `svd(a, hermitian=True)`, that does simply:
```
    if hermitian:
        # can ignore full_matrices, its a no-op here
        if compute_uv:
            s, u = eigh(a)
            return transpose(u), s, np.conjugate(u)
        else:
            return abs(eigvalsh(a))
```

File: numpy/linalg/linalg.py [hunk 1]
```diff
@@ -1947,6 +1947,82 @@ def pinv(a, rcond=1e-15):
...
+    if _isEmpty2d(a):
+        m, n = a.shape[-2:]
+        if m != n:
+            raise LinAlgError('Last 2 dimensions of the array must be square')
+        res = empty(a.shape[:-2] + (n, m), dtype=a.dtype)
+        return wrap(res)
+    s, u = eigh(a, UPLO='L')
+
+    # discard small singular values
+    cutoff = rcond[..., newaxis] * amax(s, axis=-1, keepdims=True)
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: This is actually incorrect - `eigh` can return negative values of `s`

**Comment by rgommers** [comment]
> I think the fact we allowed a hermitian overload for matrix_rank (#8557) suggests it would be reasonable to add one for the rest of linalg, at least in places where no new lapack functions are needed.

That's a good argument. Fair enough. 

> My reasoning was that Scipy uses its own internal functions, like eigh, which only operate on (M,M) shaped arrays. From what I can tell, "2d matrices only" seems to be a common theme of Scipy's linalg library, whereas Numpy provides a lot of the same functionality, but allows matrices to be stacked. I assume Scipy has a good reason for this restriction, and I don't entirely know what it is or how I would modify it if I wanted to.

I don't think there's a lot of thought put into that, it just evolved. In general, we do want APIs to match between NumPy and SciPy, and NumPy to provide the essentials and SciPy to be a superset of what NumPy has.

> In regards to the Numba issue, what would it take to add that support? I'm not super familiar with the way it interacts with Numpy/Scipy and using BLAS/LAPACK directly.

I could read the Numba code base to find the answer, but better to ask the Numba devs I think. @seibert any thoughts/comments on how Numba deals with an evolving `numpy.linalg` API here? Would it be useful to at least send Numba PRs that document that the new feature is not (yet) supported at the time we merge the new feature here?

**Comment by jordan-melendez** [comment]
@rgommers and @eric-wieser is this PR still worth pursuing given [#12693](https://github.com/numpy/numpy/pull/12693)?

**Comment by eric-wieser** [comment]
Not now that #12693 is merged. Thanks for providing the catalyst for that PR!

**Pull request closed without merging**